src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,938 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2016-2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*! DHT packet part of the toxcore.
+    * takes care of the serializing and de-serializing DHT packets
+*/
+
+use nom::{le_u8, le_u16, be_u64, be_u16};
+
+use std::net::{
+    IpAddr,
+    Ipv4Addr,
+    Ipv6Addr,
+    SocketAddr,
+    SocketAddrV4,
+    SocketAddrV6
+};
+
+use toxcore::dht_new::binary_io::*;
+use toxcore::crypto_core::*;
+
+/// Length in bytes of [`PingReq`](./struct.PingReq.html) and
+/// [`PingResp`](./struct.PingResp.html) when serialized into bytes.
+pub const PING_SIZE: usize = 9;
+
+/** `NatPing` type byte for [`NatPingReq`] and [`NatPingResp`].
+[./struct.PingReq.html] [./struct.PingResp.html]
+*/
+pub const NAT_PING_TYPE: u8 = 0xfe;
+
+/** Length in bytes of NatPings when serialized into bytes.
+*/
+pub const NAT_PING_SIZE: usize = PING_SIZE + 1;
+
+/** Standard DHT packet that encapsulates in the payload
+[`DhtPacketT`](./trait.DhtPacketT.html).
+
+https://zetok.github.io/tox-spec/#dht-packet
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DhtPacket {
+    /// [`PingReq`](./struct.PingReq.html) structure.
+    PingReq(PingReq),
+    /// [`PingResp`](./struct.PingResp.html) structure.
+    PingResp(PingResp),
+    /// [`GetNodes`](./struct.GetNodes.html) structure.
+    GetNodes(GetNodes),
+    /// [`SendNodes`](./struct.SendNodes.html) structure.
+    SendNodes(SendNodes),
+}
+
+impl ToBytes for DhtPacket {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            DhtPacket::PingReq(ref p) => p.to_bytes(buf),
+            DhtPacket::PingResp(ref p) => p.to_bytes(buf),
+            DhtPacket::GetNodes(ref p) => p.to_bytes(buf),
+            DhtPacket::SendNodes(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+impl FromBytes for DhtPacket {
+    named!(from_bytes<DhtPacket>, alt!(
+        map!(PingReq::from_bytes, DhtPacket::PingReq) |
+        map!(PingResp::from_bytes, DhtPacket::PingResp) |
+        map!(GetNodes::from_bytes, DhtPacket::GetNodes) |
+        map!(SendNodes::from_bytes, DhtPacket::SendNodes)
+    ));
+}
+
+/** DHT Request packet.
```

Comment:
I don't like the name DhtRequest. Along with DhtPacket it gives me **zero** info about what is this struct about